File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
import java.util.*;

class Solution {
        private List<String> split(String source, int length){
            List<String> tokens = new ArrayList<>();
            for(int si = 0; si < source.length(); si += length){
                int ei = si + length;
                if(ei > source.length()){
                    ei = source.length();
                }
                tokens.add(source.substring(si, ei));
                
            }
            return tokens;
        }
        
        private int compress(String source, int length){
            StringBuilder builder = new StringBuilder();
            
            String last = "";
            int count = 0;
            for(String token : split(source, length)){
                if(token.equals(last)){
                    count++;
                }
                else{
                    if(count > 1){
                        builder.append(count);
                    }
                    builder.append(last);
                        last = token;
                        count = 1;
                }
            }
            if(count > 1){
                builder.append(count);
            }
                builder.append(last);
            
                return builder.length();
            }
            
            public int solution(String s){
                int min = Integer.MAX_VALUE;
                for(int length = 1; length <= s.length(); length++){
                    int compressed = compress(s, length);
                    if(compressed < min){
                        min = compressed;
                    }
                }
                return min;
            }
        }

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.py
def solution(s):
    answer = []
    if len(s) == 1:
        return 1
    for i in range(1, len(s)+1):
        b = ''
        cnt = 1
        tmp = s[:i]
        for j in range(i,len(s)+i,i):
            if tmp == s[j:i+j]:
                cnt+=1
            else:
                if cnt != 1:
                    b += str(cnt)+tmp    
                else:
                    b += tmp
                tmp = s[j:j+i]
                cnt = 1
        answer.append(len(b))
    return min(answer)